incubator/hnc/pkg/stats/stats.go

```diff
@@ -0,0 +1,187 @@
+	hierConfigWrites counter
+
+	// namespaceWrites is the total number of Namespace writes.
+	namespaceWrites counter
+
+	objects objects
+}
+
+var stats stat
+
+// StartHierConfigReconcile updates stats when hierarchyConfig
+// reconciliation starts.
+func StartHierConfigReconcile() {
+	stats.totalHierConfigReconciles.incr()
+	stats.curHierConfigReconciles.incr()
+}
+
+// StopHierConfigReconcile updates stats when hierarchyConfig
+// reconciliation finishes.
+func StopHierConfigReconcile() {
+	stats.curHierConfigReconciles.decr()
+}
+
+// StartObjReconcile updates the stats for objects with common GK
+// when an object reconciliation starts.
+func StartObjReconcile(gk schema.GroupKind) {
+	if stats.objects == nil {
+		stats.objects = make(map[schema.GroupKind]*object)
+	}
+	if _, ok := stats.objects[gk]; !ok {
+		stats.objects[gk] = &object{
+			gk: gk,
+		}
+	}
+	stats.objects[gk].totalReconciles.incr()
+	stats.objects[gk].curReconciles.incr()
+}
+
+// StopObjReconcile updates the stats for objects with common GK
+// when an object reconciliation finishes.
+func StopObjReconcile(gk schema.GroupKind) {
+	if stats.objects == nil {
+		stats.objects = make(map[schema.GroupKind]*object)
+	}
+	if _, ok := stats.objects[gk]; !ok {
+		stats.objects[gk] = &object{
+			gk: gk,
+		}
+	}
+	stats.objects[gk].curReconciles.decr()
+}
+
+// WriteNamespace updates stats when writing namespace instance.
+func WriteNamespace() {
+	stats.namespaceWrites.incr()
+}
+
+// WriteHierConfig updates stats when writing hierarchyConfig instance.
+func WriteHierConfig() {
+	stats.hierConfigWrites.incr()
+}
+
+// WriteObject updates the object stats by GK when writing the object.
+func WriteObject(gk schema.GroupKind) {
+	stats.objects[gk].apiWrites.incr()
+}
+
+func initStats() {
+	objects := make(map[schema.GroupKind]*object)
+	for _, gvk := range config.GVKs {
+		objects[gvk.GroupKind()] = &object{
+			gk: gvk.GroupKind(),
+		}
+	}
+	stats = stat{
+		actionID: 1,
+		objects:  objects,
+	}
+}
+
+// StartLoggingActivity generates logs for performance testing.
+func StartLoggingActivity() {
+	initStats()
+	log := ctrl.Log.WithName("reconcileCounter")
+	var total, lastTotal, lastCur counter = 0, 0, 0
+	working := false
+	go func() {
+		// run forever
+		for {
+			// Log activity only when the controllers were still working in the last 0.5s.
+			time.Sleep(500 * time.Millisecond)
+			total = stats.totalHierConfigReconciles + getTotalObjReconciles()
+			// If lastCur is not 0 yet, still generate a log for the past 0.5s.
+			if total != lastTotal || lastCur != 0 {
+				// If the controller was previously idle, change its status and log it's started.
+				if working == false {
+					working = true
+					logActivity(log, "start")
+				}
+				logActivity(log, "continue")
+			} else {
+				// If the controller was previously working, change its status and log it's finished.
+				if working == true {
+					working = false
+					logActivity(log, "finish")
+					stats.actionID.incr()
```

Comment:
`actionID` doesn't _need_ to be a `counter` because it's only ever accessed from one thread. So you could just say `stats.actionID++` here if you wanted. Up to you.